Try using snackbar notices instead of prominent ones for saving/failure notices

youknowriad opened the pull request:
closes #13905, closes #11807, closes #12127, closes #15132 

This PR is a first stab towards using snackbars for post save/error and reusable blocks save/error notices.

![Screen Shot 2019-05-20 at 11 38 35](https://user-images.githubusercontent.com/699132/58012095-f8200380-7af3-11e9-922d-979d2e06bc3d.png)

Todo list:

 - [x] Improving the design
 - [x] Fix the accessibility issues (contrast and focus styles...)
 - [x] Tests
 - [x] Documentation and Changelog
 - [x] Potentially implement a timeout 

You can test quickly by creating random notices in the console:

```js
wp.data.dispatch('core/notices').createNotice(
		'info',
		'Post published.', // Text string to display.
		{
			isDismissible: true, // Whether the user can dismiss the notice.
			// Any actions the user can perform.
			type: 'snackbar',
            actions: [
				{
					onClick: () => console.log( 'test' ),
					label: 'View post'
				}
			]
		}
	);
```

youknowriad committed:
Try using snackbar notices instead of prominent ones for saving/failure notices

jasmussen commented:
Quick GIF:

![snackbars](https://user-images.githubusercontent.com/1204802/57615400-dff03780-757b-11e9-8f2d-8b5eefc1eb54.gif)

This is really nice!

Aside from a little design polish — I'll help here — this is a really great first stab. A few quick thoughts, but I'll let others chime in as well:

- Snackbars exist in addition to notices, but are great for things that are less important. Like "save successful", or "this is a reusable block, changes made affect all posts that include it", things like that. Because we can only show information that you could afford to miss, we don't really need more than a single visuals style for these, i.e. no need to distinguish levels importance with color. I would suggest we go with a dark gray.
- Snackbars disappear after a while.
- There can only be one snackbar at a time, which means if you send off multiple, they should be throttled and shown in sequence, one at a time.

nerrad commented on packages/components/src/snackbar/index.js:
```diff
@@ -0,0 +1,75 @@
+} ) {
+	const classes = classnames( className, 'components-snackbar', 'is-' + status );
+
+	if ( __unstableHTML ) {
```
How long has `__unstableHTML` been around and is this a pattern that should continue to be replicated in new things?  Is there any attempts to remove the necessity for it being __unstable?

youknowriad replied:
That's a good question, I think we should do an audit of the experimental APIs and see which ones can be made "stable".

That said, I think `__unstable` is a bit different in the sense that it's not meant to be made stable at all. It's something used internally that third-party developers shouldn't rely on. That doesn't mean the use-case here is not legit and shouldn't be exposed but it shouldn't in its current form. For this particular one @aduth knows more.

nerrad replied:
I think in this case, it was added to support back-compat with the old notice system which allowed plugins to add arbitrary html (and was related to some work @aduth did with automatically porting usage of `add_action( 'admin_notice' )` over to the notices in GB.

youknowriad replied:
Maybe we can just remove it, in that case as we don't support these admin notices anymore

nerrad replied:
I think at a minimum it should be removed for the new snackbar notices. I'm not comfortable making the call in general for all notices though.

aduth replied:
So... this one's been on my mind lately.

It's not used anywhere in Gutenberg currently. It was originally part of some compatibility effort for server-rendered notices which was dropped late in the 5.0 release cycle. In fact, I still have an open pull request to remove it: #12440

This sat longer than I'd liked, and I know of several occurrences of people indicating they're using it, despite repeated warnings of its unstable nature. The last time it came up in a #core-js chat, we said that we didn't care, and that we'd still remove it since we're not obligated to the backwards-compatibility. It doesn't sit well with me though, largely on the point that we let this specific one linger, and thus it's more "on us".

My latest thought was to do a search in the plugins repository for who might be using it, and to both (a) consider their use case and (b) reach out to them directly ahead of its removal.

I ran the search now:

https://wpdirectory.net/search/01DAVD7SA20EYCQM5K456HMYNK

There aren't quite as _many_ hits as I had expected, but certainly some popular ones.

I think regardless whether we decide to do it now (for which an argument could certainly be made), considering the use-cases is valid.

For example, one instance I'd heard that someone still chose to use `__unstableHTML` was the need for dismissed notices to be persistent. This seems like a useful feature that we'd want to consider to implement anyways.

nerrad replied:
Maybe we should flag this in #core-js and also try reaching out to those using it to gather feedback on the use-case requirements for it?  Then we could try to implement something that satisfies common use-cases.

aduth replied:
> Maybe we should flag this in #core-js and also try reaching out to those using it to gather feedback on the use-case requirements for it? Then we could try to implement something that satisfies common use-cases.

I've added it to the agenda as an item to discuss: 

https://docs.google.com/document/d/1h21-24QSi02QXr3qMqUAkzpT4B02Ep3ZKKI2qhgwUUw/edit

nerrad reviewed (commented):
I like this ui pattern so I think it'll be an improvement for certain types of notices.  A couple thoughts though: 

- In general the notices system in WordPress is in the need of an overhaul and I think GB is starting to establish new patterns for notices that _could_ eventually replace what's currently in WP.  I think some consideration needs to be given to how this works/looks for the rest of WP as a whole as we add these new patterns.
- I wonder if snackbar notices should be a bit more opinionated about what is allowed as content.  Typically they are shorter snippets visually notifying of an action that just completed.  So should there be some rules applied to not only what content is allowed in them, but also how much content?
- I wonder if notices created by WP/GB core should have a _reserved_ style for them?  Doing this would allow for easier distinction (both visually and accessibly)  between what notices are added by WordPress/GB and what were added by a plugin.

drw158 commented:
To piggyback off of Joen, some additional guidance:

- Used for low priority, non-interruptive messaging.
- Actions should **not** use a filled button, it draws too much attention.
- It doesn't have to have an action.
- They automatically disappear from the screen after a minimum of four seconds and a maximum of ten seconds.
- They shouldn't use a close icon because they disappear automatically. Although, with more research we may find that we need one.
- Don't use icons in snackbars, it draws too much attention.
- Content should be short, no more than 2 lines.
- Action button can drop to a line by itself.
- On desktop, the snackbar should be consistently placed. My preference is the bottom center, which I think works in more contexts.

youknowriad committed:
Auto-remove the snackbar notice after 10 seconds

youknowriad commented:
I added a timeout, now these notices disappear after 10 seconds.

youknowriad commented on packages/components/src/snackbar/index.js:
```diff
@@ -0,0 +1,86 @@
+} ) {
+	useEffect( () => {
+		// This rule doesn't account yet for React Hooks
+		// eslint-disable-next-line @wordpress/react-no-unsafe-timeout
```
This rule doesn't account for React Hooks. It should be possible to use and cancel timeouts in useEffect or other hooks.

youknowriad replied:
Tracked here https://github.com/WordPress/gutenberg/issues/15622

youknowriad commented on packages/components/src/snackbar/index.js:
```diff
@@ -0,0 +1,86 @@
+			tabIndex="0"
+			role="button"
+			onKeyPress={ onRemove }
+		>
```
I made the notice tabbable so you can focus it and click it to remove it but I was wondering if I should label it or is the content sufficient?

gziolo replied:
![Screen Shot 2019-05-20 at 11 07 34](https://user-images.githubusercontent.com/699132/58010616-fe60b080-7af0-11e9-8dba-e26d6d670edd.png)

This is how VoiceOver deals with it in macOS/Chrome at the moment.

youknowriad replied:
I'm thinking it would be clearer if I apply an `aria-label="__( 'Dismiss this notice' )"`? Thoughts?

youknowriad replied:
I went with that to match the current Notice component.

youknowriad commented:
> In general the notices system in WordPress is in the need of an overhaul and I think GB is starting to establish new patterns for notices that could eventually replace what's currently in WP. 

These snackbar notifications are low-priority and contextual notifications. I think they defer a little bit from what would be a notification system in WordPress where notices could be persisted across pages and have persistent actions. 

That said, the patterns introduced here can easily be reused in different pages thanks to the "notices" package.

nerrad commented on packages/components/src/snackbar/index.js:
```diff
@@ -0,0 +1,86 @@
+		}, NOTICE_TIMEOUT );
+
+		return () => clearTimeout( timeoutHandle );
+	}, [] );
```
shouldn't `onRemove` be included here as a dependency?

youknowriad replied:
Probably. Obviously I don't expect the `onRemove` prop to change over time. So the question here (and I had this in other components too is). Should we add the dependency here which could have a small performance cost (probably very small) or just don't support updating the prop?

nerrad replied:
From the react docs and what Dan Abramov has written, they've been pretty explicit in suggesting dependencies should always be included.  As for when dependencies are _known_ to not change it doesn't appear it adds much overhead.  In referencing, an example he gave with a `useEffect` listing `dispatch` (from `useReducer`) as a dependency he wrote:

> (You may omit dispatch, setState, and useRef container values from the deps because React guarantees them to be static. But it also doesn’t hurt to specify them.)

Can  you guarantee that `onRemove` will never change over time?

youknowriad replied:
I guess if it doesn't hurt :) We should add it. The question becomes, should we reset the timeout once the prop changes because this will be the default behavior if I add it there :).

nerrad replied:
well the timeout should still apply for the existing SnackNotice and it would just get reset for whatever new SnackNotice (with the changed prop) is rendered in right?  I don't think we need to worry about the reset of the timeout.

youknowriad replied:
If you add a dependency `onRemove`, the effect will trigger once the prop changes, it will clear the old timeout and trigger a new one which would reset the timeout.

youknowriad replied:
So to solve this properly we need two effects and a state :) 

 - the timeout would change a state value (shouldRemove)
 - Another effect would trigger onRemove if  "shouldRemove" is true

I'm starting to wonder if it's worth the complexity just to follow guidelines

nerrad replied:
I don't think something like this is too complex:

```js
useEffect( () => {
	// This rule doesn't account yet for React Hooks
	// eslint-disable-next-line @wordpress/react-no-unsafe-timeout
	const timeoutHandle = setTimeout( () => {
		onRemove();
	}, NOTICE_TIMEOUT );

	return () => {
            clearTimeout( timeoutHandle );
            onRemove()
        }
}, [onRemove] );
```

useEffect captures the props and state of the render call that defined it so.

youknowriad replied:
That's not the behavior we want. In your example, there are two issues:

 - If `onRemove` changes even before the timeout ends, the notice is removed (the previous `onRemove` is called)
 - if `onRemove` changes, we setup a new timeout to clear the notice after `NOTICE_TIMEOUT`ms

What we want is to clear the notice `NOTICE_TIMEOUT` after it was mounted regardless of the value of `onRemove` prop.

nerrad replied:
You're right.  I took some time to simulate what happens in various scenarios:

- with no dependency and `onRemove` prop changing
- with no dependency and `onRemove` prop not changing
- with `onRemove` as a dependency, the prop changing.
- with `onRemove` as a dependency, the prop not changing.
- with `onRemove` as a dependency, the prop changing and onRemove called with clearTimeout.

In each of the above scenarios I added an error notice and then a success notice to watch what happened it became more apparent why you set things up this way, so ya trying to to follow hooks rules here will make things more complex.

It would be good to have some tests to verify that notices are added and removed in the expected order.

sarahmonster commented:
So happy to see this happening! This is something we'll need for the navigation menu block as well, which we added in part thanks to @LukePettway's accessibility feedback. 🙌 

<img width="727" alt="Screenshot 2019-05-15 at 00 39 10" src="https://user-images.githubusercontent.com/376315/57739040-dee80480-7666-11e9-8878-c5ac29efc328.png">

Our initial mockup here uses the colour system of levels, but I'd be in agreement with @joen that these aren't strictly necessary here, especially if we create guidelines that encourage developers to avoid snackbars for critical (error) messages. Here's a quick idea of how this could potentially look:

<img width="655" alt="Screenshot 2019-05-15 at 01 34 24" src="https://user-images.githubusercontent.com/376315/57740900-edd2b500-766e-11e9-97a3-05263c684b16.png">

Let's look at how we can make this component as accessible as possible from the outset. Providing clear and consistent messaging in flexible ways has the potential to improve the product for all, by providing more helpful and contextual information without overloading users.

I've done some looking around to establish guidelines, but it would be great to get the accessibility team's input here. 

- Ensure any action is also available elsewhere in the UI so that all users can reach it. Any action link should be _supplementary_. 
- Use a Notice component for error messages, so they're persistent. (This also means the importance of colour to denote status isn't necessary, since every snackbar is then informational.)
- Use `role: status` and `aria-live: polite` to ensure the message is read out, but isn't given prominence over other, more critical elements.
- Snackbar should disappear after a relatively longer period of time to allow people time to interact with them (10s) and should also be dismissable (even if via a non-visible action, like tapping on the snackbar itself) so as to ensure it doesn't get in users' way and allow for more choice.
- Pressing the ESC key while one of the snackbar’s child elements has focus (e.g., the action button) will dismiss the snackbar.
- Copy within the snackbar should be clear and concise. (Maybe we can suggest a format here, like `verb + noun` or similar... along with an example. This is especially important when the message will only be shown for a limited time.)
- Focus should not move to the notification unless the user expressly moves to it. (Given that it's an overlay, this might present some challenges as to how exactly focus moves to the element, and there seems to be some conflicting advice on this, depending on whether the snackbar has an action or not.)

More reading:
https://inclusive-components.design/notifications/
https://www.w3.org/WAI/perspective-videos/notifications/
https://tink.uk/accessible-timeout-notifications/
https://www.lightningdesignsystem.com/components/toast/#site-main-content
https://polaris.shopify.com/components/feedback-indicators/toast#navigation

youknowriad committed:
No support for the HTML content in the snackbar notices

youknowriad committed:
Status-less snackbar notices

youknowriad committed:
Announce the notices as 'polite'

youknowriad commented on packages/notices/src/store/actions.js:
```diff
@@ -47,7 +48,11 @@ export function* createNotice( status = DEFAULT_STATUS, content, options = {} )
+			type: 'SPEAK',
+			message: content,
+			ariaLive: type === 'snackbar' ? 'polite' : 'assertive',
+		};
```
While I kept the a11y message tied to the notices package like today, I feel like a possibly better solution is to make these message as "effects" in the Notice/Snackbar components, that way you don't have to rely on the notices package for these to be announced but it would also work for the components when used directly.

I'm leaving this for a potential separate PR. cc @aduth

aduth replied:
> While I kept the a11y message tied to the notices package like today, I feel like a possibly better solution is to make these message as "effects" in the Notice/Snackbar components, that way you don't have to rely on the notices package for these to be announced but it would also work for the components when used directly.

That seems reasonable to me, though I'll open an issue to consider it for discussion and tracking: #15708

It would make sense if the spoken message would be considered as part of the rendered interface of the application, and not as much a side-effect of the data flow, which in retrospect seems the more sensible approach.

gziolo replied:
> It would make sense if the spoken message would be considered as part of the rendered interface of the application, and not as much a side-effect of the data flow, which in retrospect seems the more sensible approach.

It definitely should be something you should be able to opt in or out from. I can imagine that someone could use [ARIA live region](https://developer.mozilla.org/en-US/docs/Web/Accessibility/ARIA/ARIA_Live_Regions) instead to announce those messages.

youknowriad committed:
Add a README for the snackbar component

youknowriad committed:
Update the changelogs

youknowriad commented on packages/components/src/snackbar/README.md:
```diff
@@ -0,0 +1,48 @@
+
+1. [Design guidelines](#design-guidelines)
+2. [Development guidelines](#development-guidelines)
+3. [Related components](#related-components)
```
I think it would be a good idea to start adding "Accessibility guidelines" to all these READMEs similar to dev and design. cc @audrasjb @LukePettway

LukePettway replied:
That's a great idea!

youknowriad commented:
I updated the design, the a11y messages, I added a small README that can be iterated on, I didn't test everything yet but we can open this for review now.

youknowriad committed:
Avoid showing the error messages as snackbars

youknowriad committed:
Fix unit tests

youknowriad committed:
Fix e2e tests

gziolo commented on packages/e2e-tests/specs/reusable-blocks.test.js:
```diff
@@ -42,7 +42,7 @@ describe( 'Reusable Blocks', () => {
 
 		// Wait for creation to finish
 		await page.waitForXPath(
```
Should it be wrapped with an e2e test util to simplify usage in different places? It might be too early, but at least I can see the value of having a local function which simplifies his logic.

nerrad replied:
Ya I agree, it'd be handy to have a test util for verifying notices appear (would be good if it allows for notice type as well as content in the signature).

gziolo commented on packages/components/CHANGELOG.md:
```diff
@@ -3,6 +3,7 @@
 ### New Features
 
 - Added a new `HorizontalRule` component.
+- Added a new `Snackbar` component.
```
```suggestion
- Added a new `Snackbar` and `SnackbarList` components.
```

gziolo replied:
This wasn't fixed, was this suggestion wrong?

kjellr commented:
@youknowriad This seems like a great example of an element that we'd like to animate in and out. Animating in is super easy: we just need to assign a CSS animation. But since the element gets fully removed from the DOM when dismissed, it's a little rough to apply an animation to the removal. One way to achieve this would be to add a new class (`is-dismissed` for example) when dismissal is set to occur, and have a fade-out animation tied to that. Then, we'd trigger the actual DOM removal once the animation has completed. I'm wondering if you have any ideas for this? 

A solution to this wouldn't just be helpful here, but also for eventual subtle animations when deleting blocks, etc.

youknowriad commented:
@kjellr I think this might be super easy to achieve with https://www.react-spring.io which we're talking here and there to integrate.

kjellr commented:
> I think this might be super easy to achieve with https://www.react-spring.io which we're talking here and there to integrate.

Very cool — this demo seems promising and relevant to this PR 😄: https://codesandbox.io/embed/7mqy09jyq

youknowriad committed:
Animate the snackbar notifications

youknowriad commented:
I added fading animations @kjellr 
If there's more complexity needed, I might suggest to remove the animations and leave them out for a separate PR.

What's remaining in this PR?

LukePettway commented:
@youknowriad do we have any accessibility guidelines around animations? It's not something I see too much up but have seen it talked about before on other projects.

jasmussen commented:
> do we have any accessibility guidelines around animations? It's not something I see too much up but have seen it talked about before on other projects.

We do!

https://developer.wordpress.org/block-editor/designers/animation/

kjellr commented:
> I added fading animations @kjellr
> If there's more complexity needed, I might suggest to remove the animations and leave them out for a separate PR.

Looks great! 🎉 We can refine more in the future — I'm excited get to know react-spring more myself. 

![fade](https://user-images.githubusercontent.com/1202812/57928722-8ccdfb80-787f-11e9-8184-4b234dea1a2f.gif)

My only concern with `react-spring` so far is that I'm not sure if it's possible to opt-out of react-spring animations using the `prefers-reduced-motion` media query. It doesn't look like it from first glance.

youknowriad commented:
> My only concern with react-spring so far is that I'm not sure if it's possible to opt-out of their animations using the prefers-reduced-motion media query. It doesn't look like it from first glance.

if we were able to access the `prefers-reduced-motion` property somehow in JS, it would be possible to the check and apply the animation only when needed. That said, I'm not sure how at the moment. We could leave the animation for later if this is a hard requirement.

kjellr commented:
> if we were able to access the prefers-reduced-motion property somehow in JS, it would be possible to the check and apply the animation only when needed. That said, I'm not sure how at the moment. We could leave the animation for later if this is a hard requirement.

I'm not sure if we can do that... I haven't found anything about that in a quick search! I'll do some digging. In the meantime, let's leave the animation out for now. It's easy to add in separately. 👍

youknowriad committed:
Revert "Animate the snackbar notifications"

This reverts commit 24a058420986ac333eeb242bb243347ecedfd2aa.

youknowriad commented:
gotcha, I reverted the commit @kjellr

jasmussen commented:
Pretty sure we can do that in JS, but I refrained from commenting because my idea is hacky. Since prefers reduced is a media query, we can use a stray element as our "variable". Something like this:

```
<div class="prefers-reduced-motion" style="background: red;">
<p>I'm green if the user prefers reduced motion.</p>
</div>

<style type="text/css>
@media (prefers-reduced-motion: reduce) {
  .prefers-reduced-motion {
    background: green;
  }
}
</style>

<script type="text/javascript">
var prefersReduced = document.getElementsByClassName('prefers-reduced-motion')[0].style.background == 'green';
alert( "User prefers reduced motion: " + prefersReduced );
</script>
```

There's probably a race condition in the above, but you get my point.

kjellr commented:
[As per this helpful article](https://developers.google.com/web/updates/2019/03/prefers-reduced-motion#working_with_the_media_query), we may be able to use [`matchMedia`](https://developer.mozilla.org/en-US/docs/Web/API/Window/matchMedia) to detect the media query and then turn off the animation if it's active. 

(Hat tip to @ericwbailey for the tip 🙌)

jasmussen commented:
That's much nicer than my hack! 😂

youknowriad commented:
@kjellr This may be something we could include in the "viewport" module. @aduth I wonder if it makes more sense to rename it "device" module and include more features like this. 

But it's definitely out of scope of this PR.

youknowriad committed:
Adding the snackbar component to the docs manifest

youknowriad commented:
Can I have more reviews here? I'd like for this to be ready by the next Gutenberg release (in a week). Thanks.

nerrad commented on packages/components/src/snackbar/index.js:
```diff
@@ -0,0 +1,80 @@
+		}, NOTICE_TIMEOUT );
+
+		return () => clearTimeout( timeoutHandle );
+	}, [] );
```
Doesn't have to be done in this pull, but one thing that could get around having to use the effect rule exclusion here is to create a `useTimeout` hook that components like this can use (and the only place the es-lint rule would need  skipped is in with that custom hook definition).

youknowriad replied:
Interesting, I can see how this can be useful. Maybe we could try it once we find a second use-case?

nerrad reviewed (approved):
Did a code review and looks good to me.

gziolo commented on packages/components/src/snackbar/index.js:
```diff
@@ -0,0 +1,80 @@
+		<div
+			className={ classes }
+			onClick={ onRemove }
+			tabIndex="0"
```
Those notices are announced with `wp.a11y.speak` method like all other types of notices. They are also automatically being dismissed after 10 seconds. Do they really need to be focusable? What are the odds that keyboard-only user will want to tab to this message to dismiss it before it gets removed?

youknowriad replied:
> Pressing the ESC key while one of the snackbar’s child elements has focus (e.g., the action button) will dismiss the snackbar.

Yes, no strong opinions from me but I had this feedback earlier.

youknowriad commented on packages/components/src/snackbar/list.js:
```diff
@@ -0,0 +1,42 @@
+	return (
+		<div className={ className }>
+			{ children }
+			{ [ ...notices ].reverse().map( ( notice ) => (
```
Thinking more, I think the reverse is not needed in the snackbars. I'll just remove.

youknowriad commented on packages/components/src/snackbar/list.js:
```diff
@@ -0,0 +1,42 @@
+*/
+function SnackbarList( { notices, className, children, onRemove = noop } ) {
+	className = classnames( 'components-snackbar-list', className );
+	const removeNotice = ( id ) => () => onRemove( id );
```
yes, that's my thinking as well. If it's something that gets rerendered often (BlockEdit for instance), we should try to memoize to see if it improves performance.

For this particular component, it does seem unnecessary.

gziolo replied:
> For this particular component, it does seem unnecessary.

Yes, totally agree as mentioned in my own comment :)

gziolo commented on packages/components/src/snackbar/index.js:
```diff
@@ -0,0 +1,80 @@
+								key={ index }
+								href={ url }
+								isTertiary
+								onClick={ url ? stopPropagation : flow( stopPropagation, onClick ) }
```
Does it cover the case wher `onClick` expects `event` as its first param?

youknowriad replied:
It doesn't as written. I should update it.

youknowriad committed:
Remove the reversing of the notice list

gziolo commented on packages/editor/src/components/editor-notices/index.js:
```diff
@@ -15,17 +15,38 @@ import { compose } from '@wordpress/compose';
-	if ( dismissible !== undefined ) {
-		notices = filter( notices, { isDismissible: dismissible } );
-	}
+export function EditorNotices( { notices, onRemove } ) {
```
Nice refactor, it's much easier to follow now 👍

gziolo reviewed (commented):
I played a bit with this PR and it all looks good. Code-wise it also is looking very solid. There are some accessibility related considerations to be clarified. Let's do it before we proceed further.

One more question:

![Screen Shot 2019-05-20 at 11 38 35](https://user-images.githubusercontent.com/699132/58012095-f8200380-7af3-11e9-922d-979d2e06bc3d.png)

In the description, I see red and green backgrounds but not in the final implementation. Is this intended?

youknowriad committed:
Support snackbar onClick properly

youknowriad commented:
> In the description, I see red and green backgrounds but not in the final implementation. Is this intended?

Yes, I'll update the description. These were dismissed per design feedback.

aduth commented on packages/notices/src/store/actions.js:
```diff
@@ -38,6 +38,7 @@ export function* createNotice( status = DEFAULT_STATUS, content, options = {} )
 		context = DEFAULT_CONTEXT,
 		id = uniqueId( context ),
 		actions = [],
+		type = 'default',
```
What's the distinction between `context` and `type`? This may be an instance where the concept of a `context` was developed prematurely to its usage, but I think the general idea could be the same so far as defining a particular treatment / grouping of notices.

youknowriad replied:
I though about `context` as a way to distinguish areas of the UI. For example "editor  notices", "settings notices" where you'd display these notices in different positions of the screen. (For example right now we have block notices for image uploads, these seem to translate more to a different context for me).

I didn't see the "context" as something that would have impact on the design/persistence/usage of the notice.

That said, you're the initial implementor here, so you'd know better.

aduth replied:
> I though about `context` as a way to distinguish areas of the UI. For example "editor notices", "settings notices" where you'd display these notices in different positions of the screen. (For example right now we have block notices for image uploads, these seem to translate more to a different context for me).
> 
> I didn't see the "context" as something that would have impact on the design/persistence/usage of the notice.

The distinction becomes blurry for me based on how much we consider these things to be separate from "notices".

In my mind, `context` could make sense here if we treated them all as notices, and if NoticeList was either not prescriptive or otherwise flexible in the visual styling of individual notice items. If that were the case (and I think it could be possible?), I'd recommend `context`, then render a `NoticeList` which are styled in a way like proposed here.

If we treat it instead as an entirely separate thing, different components, then `context` might not fit as well, and we need a different taxonomy (`type`) to distinguish between a notice and a snackbar.

I guess it can also depend on whether we think that there may be _contexts_ of _snackbars_ as well as of _notices_. In which case, the dual classification may be necessary.

aduth replied:
> The distinction becomes blurry for me based on how much we consider these things to be separate from "notices".

To summarize a clarifying direct message conversation between @youknowriad and myself: From my perspective, it wasn't very clear to me in what way these, from a data perspective, were different from any other notice in the system. 

They initially seemed largely a visual treatment, which is what motivated me toward considering these as a "context" in which individual notices could be styled as a variation of the `Notice` component (e.g. `.components-notice.is-snackbar`).

As the conversation continued, initial differences were raised, including the "timing" aspect. It occurred to me whether this is something, again, distinct to "snackbar" notices, or which could be a common feature of _any_ notice.

There was finally outlined that these are quite different in _intent_, which I see now had been summarized previously at https://github.com/WordPress/gutenberg/issues/13905#issuecomment-465182984 . At that point, I see that they are in-fact different enough to warrant distinction on a `type` property. `context` can still serve for contextualizing notices within a type (e.g. global notice list vs. `withNotices` notice lists, https://github.com/WordPress/gutenberg/issues/9442#issuecomment-464436637 ).

We will still need to be careful that top-level properties of a notice (currently `content`, `isDismissible`, `speak`, `actions`) can apply to both the existing form of notice and this snackbar type. `isDismissible` is already questionable in this regard.

sarahmonster commented:
Some feedback from a visual design perspective:

- Align snackbar to middle of page, rather than left (this just looks nicer).
- Use more padding to the left and right, so padding feels more equal around the elements.
- Increase spacing between text and action link (so the distinction is super clear).
- Remove max- and min- widths, beyond a max-width of 100% (because otherwise we don't really need them).
- At mobile breakpoints, expand to full width (because otherwise, the element is likely to end up being nearly full width, which feels rather awkward).
- Make sure that when lines span full width, they don't break things.

I have a commit that can fix most of the above via CSS if it's helpful here and we'd like to implement these changes. 🚀

youknowriad commented:
Because of the fixed position of the snackbars, aligning center would be hard to achieve (account for all the variations of the WP sidebar) and also appears weird once more than one notice is shown at the same time. 

I know some initial feedback suggested that we should only show one notice at any moment but given the 10ms delay, and even if this happens, it felt better to show them at the same rather than trying to queue them.

youknowriad commented:
> Remove max- and min- widths, beyond a max-width of 100% (because otherwise we don't really need them).

I have a big screen and showing a long notice 100% width doesn't feel great either.

youknowriad commented:
Other than that, feel free to push fixes. (I can also do them if needed)

sarahmonster committed:
Improve spacing of elements inside of snackbar.

youknowriad committed:
Dismiss notice label

youknowriad committed:
Tweak the design/spacing/width of the notices

youknowriad commented:
Design and a11y feedback addressed here. Anything I'm missing? What's remaining here?

sarahmonster committed:
Ensure baselines line up properly and elements have the correct padding.

youknowriad commented on packages/components/src/snackbar/style.scss:
```diff
@@ -5,12 +5,7 @@
-
-	@include break-small() {
-		width: fit-content;
-	}
```
Any reason to remove the "fit-content" ?

This causes this 
<img width="789" alt="Capture d’écran 2019-05-22 à 1 32 41 PM" src="https://user-images.githubusercontent.com/272444/58174610-31907480-7c96-11e9-99ec-ca7159b6cd8d.png">

The first notice takes the width of the second one.

sarahmonster replied:
I'm mucking around trying to get centred alignments working and getting a bit lost in the multiple different file locations, so there may be some 🚧 in my commits here until it's all ironed out.

My _hope_ is that there'd never be an instance where two snackbar notifications appear at once, so this would be an edge case. (We can add this to the design guidelines.) But in the case that two *do* show up, it's probably better if they're the same width, otherwise it would feel super jarring.

sarahmonster committed:
On mobile screens, show full-width snackbar.

sarahmonster committed:
Attempt to centre snackbar notice within editing window.

sarahmonster committed:
Ensure that multiple snackbars stack well.

sarahmonster committed:
Reduce max-width for better legibility of long lines.

youknowriad commented on packages/editor/src/components/editor-notices/style.scss:
```diff
@@ -34,5 +34,20 @@
+
+	@include break-large() {
+		width: calc(100% - #{ $admin-sidebar-width });
+	}
```
Ideall,y the styles that are specific to "WP Admin" should move to this stylesheet `packages/edit-post/src/components/layout/style.scss` since the styles in this file can be used in any context (think a page where there's not sidebar...)

sarahmonster replied:
👍 That's really helpful to know—I was wondering why there were styles for this component outside of the component itself. It makes things a bit confusing to work with. I'll try to move anything that refers to the sidebar over to that file.

youknowriad commented on packages/edit-post/src/components/layout/style.scss:
```diff
@@ -30,7 +30,8 @@
 @include editor-left(".edit-post-layout__content .components-editor-notices__snackbar");
```
I'm assuming this mixin is not needed anymore if you want to center them

sarahmonster replied:
👍 Definitely had no clue what this was for. 😜

sarahmonster committed:
Move positioning styles to layout file and nix unneeded styles.

sarahmonster commented:
I've done my best to fix the design issues outlined [my comment above](https://github.com/WordPress/gutenberg/pull/15594#issuecomment-494177568) with the exception of the animating-in issue, which we can look into in a separate PR.

I had a bit of troubles with getting the notices centred on all permutations of the sidebar view. I think it may be that I'm not fully grokking how the `auto-fold` and `folded` CSS classes are applied to the document body, and I'm missing something. I couldn't figure out the proper CSS class to target in these situations:

<img width="644" alt="Screenshot 2019-05-22 at 14 35 42" src="https://user-images.githubusercontent.com/376315/58243125-fd758c00-7d47-11e9-8eb3-5fe5563f831e.png">

<img width="413" alt="Screenshot 2019-05-22 at 14 35 56" src="https://user-images.githubusercontent.com/376315/58243126-fd758c00-7d47-11e9-83fd-2adb5af5c467.png">

<img width="718" alt="Screenshot 2019-05-22 at 14 46 31" src="https://user-images.githubusercontent.com/376315/58243095-f0f13380-7d47-11e9-8e85-00c534d4b024.png">

For what it's worth, the second scenario here suffers the same issues with the left-aligned approach, so it's probably only critical that we fix the other two scenarios. I'd love some guidance here from someone who's a bit more familiar with the CSS code than I am, since I'm finding it a bit of a challenge to wade through. 🤯

youknowriad commented:
> I had a bit of troubles with getting the notices centred on all permutations of the sidebar view. I think it may be that I'm not fully grokking how the auto-fold and folded CSS classes are applied to the document body

That's exactly why I suggested that we keep the left aligned ones. I also prefer them visually (but I can see how this is just a personal difference). Maybe @jasmussen could help here

jasmussen commented:
> I had a bit of troubles with getting the notices centred on all permutations of the sidebar view. I think it may be that I'm not fully grokking how the auto-fold and folded CSS classes are applied to the document body, and I'm missing something. I couldn't figure out the proper CSS class to target in these situations:

Yes, this is quite messy. We have some mixins to take some of the vinegar out of here, but even then it's not superb.

In fact although it may look nice when centered according to the main column, perhaps it would be worth to think of these notices outside of the boundaries of the editing canvas, but even think of them more as "global notices" for all of WordPress. Which means we could simply center them absolutely, regardless of left or right sidebars (easiest), or even left align them on desktop breakpoints like Gmail does:

<img width="939" alt="Screenshot 2019-05-23 at 11 59 04" src="https://user-images.githubusercontent.com/1204802/58244205-737ef080-7d52-11e9-847b-6a97a1f8dfaf.png">

The [material guidelines](https://material.io/design/components/snackbars.html#placement) seem to suggest the same, which we can choose to take inspiration from.

aduth commented:
Would this close #15132 ?

youknowriad committed:
Restore left aligned snackbars

youknowriad commented:
In an attempt to push this into a mergeable state and given the technical complexities and the fact that the left aligned snackbars can work just as well. I restored the left alignment for now.

What's left in this PR? Can I get an approval?

jasmussen commented:
From a design POV, 👍 👍. We can always explore additional tweaks to how to position this best in any subsequent PRs, but it shouldn't block a wonderful component from shipping. All this needs is a code sanity check.

youknowriad commented:
Ok let's ship this @aduth and @nerrad already reviewed this from a code perspective and it's good to land early in the release timeline.

mapk commented:
🎉 Excellent work! Thanks for exploring this @youknowriad!

jwold commented:
Just want to add my two cents, that from a design perspective the snackbar notice feels much more elegant and simple. Love it!

QuietNoise commented:
They are lovely. 
Can I comment on the fact that they are however not UX perfect when creating new posts. 
Because they disappear over time there is no quick way to 'View post' because on new posts the View post link does not yet appear in Admin bar (your example shows that as well).

Admin has to either re-save to see the notice again or refresh te page to see the View psot link in admin bar.

dingo-d commented:
Why aren't these called toaster notices? Like in every other place on the web? You even linked to an article on salesforce that mentions them as `toast` notices... Nobody knows what the 'snackbar' is because they haven't heard of it... Why are you confusing people? 😕

jasmussen commented:
> Because they disappear over time there is no quick way to 'View post' because on new posts the View post link does not yet appear in Admin bar (your example shows that as well).

This is a good point. In absence of a different interface for viewing post, perhaps that notification should be a dismissible notice instead of a snackbar. Would you mind opening a new ticket for this? Feel free to CC me.

youknowriad commented:
I think I disagree with this, when you publish a post, the "view link" is always visible on the post-publish panel unless you start typing.

If we want to improve further, I'd rather think of ways to show an always visible "View post" link in the UI instead of relying on notices.

sarahmonster commented:
One of our design principles for the snackbar notifications is that any action link will always be available elsewhere in the UI. 

This link isn't super prominent right now though, and (I suspect) that the most common action users want to take after publishing a post is viewing that post. IReverting back to a more disruptive notice may not be the best solution here—instead, we should either provide a more prominent link in the post-publish interface or consider redirecting the user to the post, and showing a notification with more post-publish options. (See: changes suggested in #7602.)
